libdnf5/module/module_item.cpp

```diff
@@ -38,58 +38,149 @@ extern "C" {
 
 #include <algorithm>
 #include <string>
+#include <utility>
 #include <vector>
 
 namespace libdnf5::module {
 
+class ModuleItem::Impl {
+public:
+    Impl(_ModulemdModuleStream * md_stream, const ModuleSackWeakPtr & module_sack, std::string repo_id);
+
+    ~Impl();
+    Impl(const Impl & mpkg);
+    Impl & operator=(const Impl & mpkg);
+    Impl(Impl && mpkg) noexcept;
+    Impl & operator=(Impl && mpkg) noexcept;
+
+private:
+    friend ModuleItem;
+
+    ModuleSackWeakPtr module_sack;
+
+    ModuleItemId id;
+    std::string repo_id;
+
+    // Corresponds to one yaml document
+    _ModulemdModuleStream * md_stream;
+
+    // For compatibility with older modules that didn't have static contexts
+    std::string computed_static_context;
+};
+
+ModuleItem::Impl::Impl(_ModulemdModuleStream * md_stream, const ModuleSackWeakPtr & module_sack, std::string repo_id)
+    : module_sack(module_sack),
+      repo_id(std::move(repo_id)),
+      md_stream(md_stream) {
+    if (md_stream != nullptr) {
+        g_object_ref(md_stream);
+    }
+}
+
+
+ModuleItem::Impl::~Impl() {
+    if (md_stream != nullptr) {
+        g_object_unref(md_stream);
+    }
+}
+
+ModuleItem::Impl::Impl(const Impl & mpkg)
+    : module_sack(mpkg.module_sack),
+      id(mpkg.id),
+      repo_id(mpkg.repo_id),
+      md_stream(mpkg.md_stream),
+      computed_static_context(mpkg.computed_static_context) {
+    if (md_stream != nullptr) {
+        g_object_ref(md_stream);
+    }
+}
+
+
+ModuleItem::Impl & ModuleItem::Impl::operator=(const Impl & mpkg) {
+    if (this != &mpkg) {
+        module_sack = mpkg.module_sack;
+        id = mpkg.id;
+        repo_id = mpkg.repo_id;
+        md_stream = mpkg.md_stream;
+        if (md_stream != nullptr) {
+            g_object_ref(md_stream);
+        }
+        computed_static_context = mpkg.computed_static_context;
+    }
+    return *this;
+}
+
+
+ModuleItem::Impl::Impl(Impl && mpkg) noexcept
+    : module_sack(mpkg.module_sack),
+      id(mpkg.id),
+      repo_id(mpkg.repo_id),
+      md_stream(mpkg.md_stream),
+      computed_static_context(mpkg.computed_static_context) {
+    mpkg.md_stream = nullptr;
+}
+
+
+ModuleItem::Impl & ModuleItem::Impl::operator=(Impl && mpkg) noexcept {
+    if (this != &mpkg) {
+        module_sack = mpkg.module_sack;
+        id = mpkg.id;
+        repo_id = mpkg.repo_id;
+        md_stream = mpkg.md_stream;
+        mpkg.md_stream = nullptr;
+        computed_static_context = mpkg.computed_static_context;
+    }
+    return *this;
+}
```

Comment:
Hi, probably a stupid question, but in the previous code, `std::move` was used for some of the attributes. Is it for some reason done automatically in this case? If so, can it be relied on? Or is there some other reason for removing it?

Reply:
Oh, I have added it back for the strings (`computed_static_context` and `repo_id`), thanks.
Regarding `id` it is trivially-copyable type (it is basically an `int`) so the move had not effect.
For `module_sack` it also had no effect because `ModuleSackWeakPtr` doesn't have enabled move constructors.